k2/csrc/cuda/ops.h

```diff
@@ -1,71 +1,142 @@
 // k2/csrc/cuda/ops.h
 
-// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey)
+// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey
+//                                                   Haowen Qiu)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_CUDA_OPS_H_
 #define K2_CSRC_CUDA_OPS_H_
 
-#include "k2/csrc/cuda/array.h"
+#include <cassert>
+#include <type_traits>
 
+#include "k2/csrc/cuda/array.h"
 
-// Note, I'm not sure about the name of this file, they are not ops like in TensorFlow, but
-// procedures..
+// Note, I'm not sure about the name of this file, they are not ops like in
+// TensorFlow, but procedures..
+
+// TODO(haowen): this function may be moved to some header files for log (e.g.
+// `log.h`)
+inline cudaError_t CheckCuda(cudaError_t status) {
+#if defined(DEBUG)
+  if (status != cudaSuccess) {
+    fprintf(stderr, "CUDA Runtime Error: %s\n", cudaGetErrorString(status));
+    assert(status == cudaSuccess);
+  }
+#endif
+  return status;
+}
+
+// TODO(haowen): manage/load block config with some classes? then we can get
+// different configuration depending on num_elements and data type.
+constexpr int TILE_DIM = 32;
+constexpr int BLOCK_ROWS = 8;
```

Comment:
My feeling is that it would make more sense to not make this fixed... in many circumstances, when transposing small matrices, having one kernel process one element would likely make more sense.  Unless you know something I don't.

Reply:
so do you suggest we should check the number of elements do one-by-one transposing for small matrices? Using shared memory without bank conflicts usually will get more than 5x bandwidth here.

Reply:
BTW, Using a thread block with fewer threads than elements here will lead that much of the index calculation cost is amortized over these elements. I will run tests locally to see the gain and may remove the loop if the gain is not so much.

Reply:
@danpovey Dan, I tested the matrix transpose on xiaomi-cluster (V100), seems amortized index calculation cost can get some gain, without loop (one thread one element), the effective  bandwidth is about 630GB/s, with loop (one thread processes 4 elements), the effective bandwidth is about 720GB/s.

Reply:
will keep the for-loop here for now as the bandwidth gain.